google/cloud/pubsublite/internal/futures.h

```diff
@@ -0,0 +1,51 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_PUBSUBLITE_INTERNAL_FUTURES_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_PUBSUBLITE_INTERNAL_FUTURES_H
+
+#include "google/cloud/future.h"
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace pubsublite_internal {
+
+template <class T>
+struct ChainFutureImpl {
+  template <class U>
+  // Note: this drops any exceptions contained in `future<U>`
+  future<T> operator()(future<U>) {
+    return std::move(f);
+  }
+
+  future<T> f;
+};
+
+/**
+ * This is a templatized helper to chain futures together with an intended
```

Comment:
Well, yes. I can see that.  My question is why do you need this instead of some `.then()` calls and lambdas? Is that because you need to capture-by-move the returned future? That is, because this does not work:

```
future<T> f  = ...;
future<U> g = ...;
return g.then([ff = std::move(f)](future<U>) mutable { return std::move(ff); }
```

If so, maybe add something like "We often need to capture and return futures from `.then()` continuations, but without C++14 support, this is very tedious, this class and function reduce the effort" to the comment.

Reply:
+1, this is the reason for this helper.

Reply:
Can we change the comments to say that?  Maybe something along these lines:

```cc
/**
  * A helper to capture-by-move futures into a second future continuation.
  *
  * Given two futures `future<U> r` and `future<T> f` we often want to write:
   * @code
   * f.then([tmp = std::move(f)](future<T>) mutable { return std::move(tmp); } 
   * @encode
  *
  * Unfortunately we cannot, as the project needs to support C++11. This is 
  * a helper to avoid repetition on this pattern.
  */
```